### ISSUE TITLE
Regression in cloud_firestore and firebase_database Android Transactions

### ISSUE THREAD

**Issue by collinjackson** [flutter/flutter#32657]
I'm noticing a regression in the transactions integration test on Android on the latest Flutter master. The test is passing on Flutter stable.

Command to run test:
```
flutter drive test_driver/cloud_firestore.dart 
```

On Flutter stable:
```
I/flutter ( 6535): Observatory listening on http://127.0.0.1:34300/tsSx9QLxaFA=/
[info ] FlutterDriver: Connecting to Flutter application at http://127.0.0.1:61653/tsSx9QLxaFA=/
[trace] FlutterDriver: Isolate found with number: 617548898
[trace] FlutterDriver: Isolate is paused at start.
[trace] FlutterDriver: Attempting to resume isolate
[trace] FlutterDriver: Waiting for service extension
I/flutter ( 6535): 00:00 +0: Firestore getDocuments
[info ] FlutterDriver: Connected to Flutter application.
I/flutter ( 6535): 00:01 +1: Firestore increment
I/flutter ( 6535): 00:03 +2: Firestore runTransaction
I/flutter ( 6535): 00:04 +3: Firestore pagination
I/flutter ( 6535): 00:05 +4: (tearDownAll)
I/flutter ( 6535): 00:05 +5: All tests passed!
```

On Flutter master:
```
[info ] FlutterDriver: Connected to Flutter application.
I/flutter ( 9558): 00:00 +0: Firestore getDocuments
I/flutter ( 9558): 00:01 +1: Firestore increment
I/flutter ( 9558): 00:02 +2: Firestore runTransaction
I/flutter ( 9558): 00:32 +2: Firestore runTransaction [E]
I/flutter ( 9558):   TimeoutException after 0:00:30.000000: Test timed out after 30 seconds.
I/flutter ( 9558):   package:test_api/src/backend/invoker.dart 332:28              Invoker._handleError.<fn>
I/flutter ( 9558):   dart:async/zone.dart 1120:38                                  _rootRun
I/flutter ( 9558):   dart:async/zone.dart 1021:19                                  _CustomZone.run
I/flutter ( 9558):   package:test_api/src/backend/invoker.dart 330:10              Invoker._handleError
I/flutter ( 9558):   package:test_api/src/backend/invoker.dart 283:9               Invoker.heartbeat.<fn>.<fn>
I/flutter ( 9558):   dart:async/zone.dart 1124:13                                  _rootRun
I/flutter ( 9558):   dart:async/zone.dart 1021:19                                  _CustomZone.run
I/flutter ( 9558):   package:test_api/src/backend/invoker.dart 281:38              Invoker.heartbeat.<fn>
I/flutter ( 9558):   package:stack_trace/src/stack_zone_specification.dart 209:15  StackZoneSpecification._run
I/flutter ( 9558):   package:stack_trace/src/stack_zone_specification.dart 119:48  StackZoneSpecification._registerCallback.<fn>
I/flutter ( 9558):   dart:async/zone.dart 1124:13                                  _rootRun
I/flutter ( 9558):   dart:async/zone.dart 1021:19                                  _CustomZone.run
I/flutter ( 9558):   dart:a
I/flutter ( 9558): 00:32 +2 -1: Firestore pagination
I/flutter ( 9558): 00:34 +3 -1: (tearDownAll)
I/flutter ( 9558): 00:34 +4 -1: Some tests failed.
```

The failure seems repeatable rather than a one-time flake.

It's possible that a Flutter platform change is interfering with the Android transactions implementation.

**Comment by collinjackson** [comment]
`git bisect` narrowed down the Flutter revision where the regression occurred to c42f3fb1d318a74ea186439f3f5083f56cca44f5 which corresponds to a roll to 2c9e37c34e79. It seems most likely that the regression is flutter/engine#8830 which causes a crash in FlutterJNI if invoked on a non-main thread in debug mode. And sure enough we are hitting the crash in the adb logs (not showing up on the test logs unfortunately).

```
firestoreexample, PID: 21909
05-13 18:12:30.297 21909 21973 E AndroidRuntime: java.lang.RuntimeException: An error occurred while executing doInBackground()
05-13 18:12:30.297 21909 21973 E AndroidRuntime: 	at android.os.AsyncTask$3.done(AsyncTask.java:354)
05-13 18:12:30.297 21909 21973 E AndroidRuntime: 	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
05-13 18:12:30.297 21909 21973 E AndroidRuntime: 	at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
05-13 18:12:30.297 21909 21973 E AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:271)
05-13 18:12:30.297 21909 21973 E AndroidRuntime: 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
05-13 18:12:30.297 21909 21973 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
05-13 18:12:30.297 21909 21973 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
05-13 18:12:30.297 21909 21973 E AndroidRuntime: 	at java.lang.Thread.run(Thread.java:764)
05-13 18:12:30.297 21909 21973 E AndroidRuntime: Caused by: java.lang.RuntimeException: Methods marked with @UiThread must be executed on the main thread. Current thread: AsyncTask #3
05-13 18:12:30.297 21909 21973 E AndroidRuntime: 	at io.flutter.embedding.engine.FlutterJNI.ensureRunningOnMainThread(FlutterJNI.java:605)
05-13 18:12:30.297 21909 21973 E AndroidRuntime: 	at io.flutter.embedding.engine.FlutterJNI.invokePlatformMessageResponseCallback(FlutterJNI.java:556)
05-13 18:12:30.297 21909 21973 E AndroidRuntime: 	at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(DartMessenger.java:131)
05-13 18:12:30.297 21909 21973 E AndroidRuntime: 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler$1.success(MethodChannel.java:225)
05-13 18:12:30.297 21909 21973 E AndroidRuntime: 	at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$5.doInBackground(CloudFirestorePlugin.java:459)
05-13 18:12:30.297 21909 21973 E AndroidRuntime: 	at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$5.doInBackground(CloudFirestorePlugin.java:452)
05-13 18:12:30.297 21909 21973 E AndroidRuntime: 	at android.os.AsyncTask$2.call(AsyncTask.java:333)
05-13 18:12:30.297 21909 21973 E AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
05-13 18:12:30.297 21909 21973 E AndroidRuntime: 	... 4 more
```

It may be possible to work around this issue by moving the JNI calls to the correct thread, I'm investigating.

@amirh @dnfield

**Comment by elmarou** [comment]
I want to use the authentication registration feature with a Google Account, but when the (_singInWithGoogle) function is called, the application stops and displays the following error.

````
D/AndroidRuntime(11262): Shutting down VM 
E/AndroidRuntime(11262): FATAL EXCEPTION: main 
E/AndroidRuntime(11262): Process: com.hayriyad.erpstage, PID: 11262 E/AndroidRuntime(11262): java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=53293, result=0, data=null} to activity {com.hayriyad.erpstage/com.hayriyad.erpstage.MainActivity}: java.lang.IllegalStateException: Reply already submitted E/AndroidRuntime(11262): at android.app.ActivityThread.deliverResults(ActivityThread.java:4360) E/AndroidRuntime(11262): at android.app.ActivityThread.handleSendResult(ActivityThread.java:4402) E/AndroidRuntime(11262): at android.app.servertransaction.ActivityResultItem.execute(ActivityResultItem.java:49) E/AndroidRuntime(11262): 
```

2.***i use this fonction*** 
```dart
void _signInWithGoogle() async {
    final GoogleSignInAccount googleUser = await _googleSignIn.signIn();
    final GoogleSignInAuthentication googleAuth =
    await googleUser.authentication;
    final AuthCredential credential = GoogleAuthProvider.getCredential(
      accessToken: googleAuth.accessToken,
      idToken: googleAuth.idToken,

    );
    final FirebaseUser user = await _auth.signInWithCredential(credential);
    assert(user.email != null );
    assert(user.displayName != null);
    assert(!user.isAnonymous);
    assert(await user.getIdToken() != null);

    final FirebaseUser currentUser = await _auth.currentUser();
    assert(user.uid == currentUser.uid);
    setState(() {
      if (user != null) {
        _success = true;
        _userID = user.uid;
      } else {
        _success = false;
      }
    });
  }```

**Comment by collinjackson** [comment]
@elmarou I think that your question is unrelated to this issue.

**Comment by kentcb** [comment]
On Android, my transaction logic jumps into the `Firestore.runTransaction` method, which kicks off the `channel.invokeMethod` call, but that never returns. I don't see anything in logcat or debug output. Trying to get Android Studio to cooperate so I can debug the native side.

Works fine on iOS.

Is this the same issue? I noticed that https://github.com/flutter/engine/pull/8830 is merged and I updated to latest `master` to check, but I still get the problem.
